### PR TITLE
[SPARK-33456][SQL][TEST] Add end-to-end test for subexpression elimination

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to add end-to-end test for subexpression elimination.

### Why are the changes needed?

We have subexpression elimination feature for expression evaluation but we don't have end-to-end tests for the feature. We should have one to make sure we don't break it.

### Does this PR introduce _any_ user-facing change?

No, dev only.

### How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun @maropu @cloud-fan

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @viirya .

**Review by dongjoon-hyun — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by viirya** [comment]
Thanks @HyukjinKwon @dongjoon-hyun

File: sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+--SET spark.sql.optimizer.enableJsonExpressionOptimization=false
+
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=true
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=false
+
+--CONFIG_DIM2 spark.sql.codegen.factoryMode=CODEGEN_ONLY
+--CONFIG_DIM2 spark.sql.codegen.factoryMode=NO_CODEGEN
+
+--CONFIG_DIM3 SUBEXPRESSION_ELIMINATION_ENABLED=true
+--CONFIG_DIM3 SUBEXPRESSION_ELIMINATION_ENABLED=false
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql:12]: Wait, shouldn't it be the actual config names?

File: sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+--SET spark.sql.optimizer.enableJsonExpressionOptimization=false
+
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=true
+--CONFIG_DIM1 spark.sql.codegen.wholeStage=false
+
+--CONFIG_DIM2 spark.sql.codegen.factoryMode=CODEGEN_ONLY
+--CONFIG_DIM2 spark.sql.codegen.factoryMode=NO_CODEGEN
+
+--CONFIG_DIM3 SUBEXPRESSION_ELIMINATION_ENABLED=true
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql:11]: is `SUBEXPRESSION_ELIMINATION_ENABLED` a config name?

File: sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql (hunk 1 above)
Comment (viirya) [reply sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql:12]: Oops, let me fix it.

Comment (viirya) [reply sql/core/src/test/resources/sql-tests/inputs/subexp-elimination.sql:12]: Fix in #30384.

**Comment by maropu** [comment]
Thanks for adding it, @viirya !

**Pull request closed without merging**